Add Camera2D Rotation Smoothing

luis-l opened the pull request:
`Camera2D` has follow smoothing to interpolate towards a target position, but no rotation smoothing to align with the target rotation. 

This adds rotation smoothing directly into the `Camera2D` API by  having two new properties: 
- `smoothing_rotation_enabled`
- `smoothing_rotation_speed`

Possible issue: confusion by having two similar properties show in the editor: `Rotating` and` Smoothing/Rotation Enabled`. I am open to suggestions to improve this. 

Note that rotational smoothing only works if `Rotating` is enabled.

![image](https://user-images.githubusercontent.com/7874807/80435111-e7a60a80-88b8-11ea-8afb-c228b02a9b16.png)

[Camera2D Smooth Rotation Video](https://www.reddit.com/r/godot/comments/g8ov6q/got_rotation_smoothing_built_directly_into/?utm_source=share&utm_medium=web2x)

mhilbrunner commented:
Note: this PR seems to inclue the commit from https://github.com/godotengine/godot/issues/38276 too

luis-l commented:
> 
> 
> Note: this PR seems to include the commit from #38276 too

I could not find a way to do a PR dependency or find a preferred way to handle this.

Calinou commented:
@luis-l GitHub doesn't offer a way to mark PRs as dependent, so you'll just need to write in your PR description that this PR should be merged *after* #38278.

luis-l committed:
Add rotation smoothing to Camera2D

Camera2D has follow smoothing to interpolate towards a
target position, but no rotation smoothing to align
with the target rotation.

This adds rotation smoothing directly into the Camera2D API
by having two new properties:
  - smoothing_rotation_enabled
  - smoothing_rotation_speed

luis-l commented:
Camera2D smoothing rotation now uses lerp_angle. **No longer depends on** #38276 (which has been closed)

bojidar-bg commented on scene/2d/camera_2d.cpp:
```diff
@@ -754,6 +794,8 @@ void Camera2D::_bind_methods() {
 	ADD_GROUP("Smoothing", "smoothing_");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "smoothing_enabled"), "set_enable_follow_smoothing", "is_follow_smoothing_enabled");
 	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "smoothing_speed"), "set_follow_smoothing", "get_follow_smoothing");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "smoothing_rotation_enabled"), "set_enable_rotation_smoothing", "is_rotation_smoothing_enabled");
```
To reduce the confusion mentioned in the OP, maybe move those two to a group called "Rotation Smoothing"?

(Also: `smoothing_rotation_enabled` should toggle internal process too, if it is to be usable without `Smoothing` enabled)

luis-l replied:
Another group for Rotation Smoothing would work. 

Should the "Smoothing" group be renamed to "Follow Smoothing" (or something similar) to help clarify the distinction between the two smooth groups?

luis-l replied:
> (Also: `smoothing_rotation_enabled` should toggle internal process too, if it is to be usable without `smoothing_enabled`)

What will be the best way to handle internal_process toggling when one smoothing is > 0 and the other is not?

For example if follow smoothing is > 0 but rotation smoothing is <= 0, is it valid to set internal process to true?

bojidar-bg replied:
If either smoothing or rotation smoothing is enabled, you need process on. So, doesn't matter which one of the two requires it, just one of them is enough.

luis-l replied:
@bojidar-bg 
> If either smoothing or rotation smoothing is enabled, you need process on. So, doesn't matter which one of the two requires it, just one of them is enough.

Could you clarify why we enable internal process for the given condition
```
if (smoothing > 0 && !(is_inside_tree() && Engine::get_singleton()->is_editor_hint()))
      set_process_internal(true);
```
So it can be set true only smoothing has a value and is not inside the tree and not in editor.

luis-l commented:
Two groups for smoothing: Follow and Rotation
![image](https://user-images.githubusercontent.com/7874807/80630355-4f6f6900-8a11-11ea-8b3f-17d402db8c93.png)

luis-l commented:
Now that I think of it, Follow Smoothing should just be Smoothing so it matches with the variable names: smoothing_enabled and smoothing_speed.

clayjohn commented:
Reviewed briefly in PR review meeting, seems fine, needs detailed review from someone more familiar with the 2D engine, e.g. @KoBeWi

KoBeWi commented:
It should be rebased first.

KoBeWi commented:
Superseded by #65776
